keps/sig-node/4191-split-image-filesystem/README.md

```diff
@@ -0,0 +1,899 @@
+    repeated FilesystemUsage container_filesystems = 2;
+}
+
+// FilesystemUsage provides the filesystem usage information.
+message FilesystemUsage {
+    // Timestamp in nanoseconds at which the information were collected. Must be > 0.
+    int64 timestamp = 1;
+    // The unique identifier of the filesystem.
+    FilesystemIdentifier fs_id = 2;
+    // UsedBytes represents the bytes used for images on the filesystem.
+    // This may differ from the total bytes used on the filesystem and may not
+    // equal CapacityBytes - AvailableBytes.
+    UInt64Value used_bytes = 3;
+    // InodesUsed represents the inodes used by the images.
+    // This may not equal InodesCapacity - InodesAvailable because the underlying
+    // filesystem may also be used for purposes other than storing images.
+    UInt64Value inodes_used = 4;
+}
+```
+
+CRIO returns a unique identifier for images and containers so Kubelet can ask CRI if the objects are stored on separate disks.  
+In the dedicated disk for container runtime, images_filesystem and container_fileystem will be set to the same value.
+
+CRI implementation can set this as needed.  The image and container filesystems are both arrays so this provides some extensibility in case these are stored on multiple disks.  
+
+Container runtimes will need to implement ImageFsInfo
+
+- [Crio Implementation](https://github.com/cri-o/cri-o/blob/main/server/image_fs_info.go)
+- [Containerd implementation](https://github.com/containerd/containerd/blob/main/pkg/cri/server/imagefs_info.go)
+
+An Alpha goal would be to have an implementation of `crictl imagefsinfo` that can allow for more detailed reports of the image fs info.
+
+See [PR](https://github.com/kannon92/cri-tools/pull/1) for an example.
+
+### Stats Summary
+
+Stats Summary has a field called runtime and we will add a containerFs to the runtime field.
+
+```golang
+// NodeStats holds node-level unprocessed sample stats.
+type NodeStats struct {
+  ...
+// Reference to the measured Node.
+// Stats about the underlying container runtime.
+// +optional
+Runtime *RuntimeStats `json:"runtime,omitempty"`
+}
+
+// RuntimeStats are stats pertaining to the underlying container runtime.
+type RuntimeStats struct {
+// Stats about the underlying filesystem where container images are stored.
+// This filesystem could be the same as the primary (root) filesystem.
+// Usage here refers to the total number of bytes occupied by images on the filesystem.
+// +optional
+ImageFs *FsStats `json:"imageFs,omitempty"`
+// Stats about the underlying filesystem where container's writeable layer is stored.
+// This filesystem could be the same as the primary (root) filesystem or the ImageFS.
+// Usage here refers to the total number of bytes occupied by the writeable layer on the filesystem.
+// +optional
+ContainerFs *FsStats `json:"containerFs,omitempty"`
+}
+```
+
+In this KEP, ContainerFs can either be the same as ImageFs or NodeFs.
```

Comment:
should it always be reported? how should kubelet / eviction manager act if ImageFs is specified but ContainerFs isn't (for instance, how will newer kubelet handle older containerd?)

Reply:
I will guard against that case.  I'll mention that if the container runtime does not support this, then we will assume ImageFs=ContainerFs.